### PR TITLE
Fix #4447 - Teach MVC to understand defaults

### LINKED ISSUE #4447: MVC not honoring attribute route default values for URL matching
See: https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs#L14

MVC isn't propagating the default values from a URL template. That means that code like:
```
        [HttpGet("/TeamName/{*Name=DefaultName}/")]
        public ActionResult GetTeam(string name)
        {
        }
```

will match `/TeamName` but will always have a default value of `null` for `name`

The fix here to pass `routeInfo.Defaults` + the secret route token.

### PR DESCRIPTION
**Pull request by rynowak**
The wireup for defaults in attribute routes was missing

### CODE REVIEW

**Comment by rynowak** [comment]
FYI this is an RC2 release bug - ignore the fact that I based the PR on dev.

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Abstractions/ActionDescriptor.cs [hunk 1]
```diff
@@ -53,6 +53,6 @@ public ActionDescriptor()
         /// <summary>
         /// Stores arbitrary metadata properties associated with the <see cref="ActionDescriptor"/>.
         /// </summary>
-        public IDictionary<object, object> Properties { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Abstractions/ActionDescriptor.cs]: Made these get/set for unit testing. It's consistent with everything else in this class

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs [hunk 2]
```diff
@@ -133,24 +138,44 @@ private TreeRouter BuildRoute(ActionDescriptorCollection actions)
...
             foreach (var routeInfo in distinctRouteInfosByGroup)
             {
-                routeBuilder.Add(new TreeRouteMatchingEntry()
+                // Note that because we only support 'inline' defaults, each routeInfo group also has the same
+                // set of defaults. 
+                //
+                // We then inject the route group as a default for the matcher so it gets passed back to MVC
+                // for use in action selection.
+                var defaults = new RouteValueDictionary(routeInfo.Defaults);
+                defaults[TreeRouter.RouteGroupKey] = routeInfo.RouteGroup;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs]: This is the only new code. The rest is just a refactor for testability

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/AttributeRouteTest.cs [hunk 3]
```diff
@@ -118,5 +103,513 @@ public class AttributeRouteTest
...
+                    Assert.Same(handler, e.Target);
+                    Assert.Collection(
+                        e.TemplateMatcher.Defaults.OrderBy(kvp => kvp.Key),
+                        kvp => Assert.Equal(new KeyValuePair<string, object>(TreeRouter.RouteGroupKey, "1"), kvp));
+                    Assert.Same(e.RouteTemplate, e.TemplateMatcher.Template);
+                });
+        }
+
+        [Fact]
+        public void AttributeRoute_GetEntries_CreatesMatchingEntry_WithDefault()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/AttributeRouteTest.cs]: This is the test that verifies the fix (along with the tweaked functional test). There are tests for defaults in attribute routes, but they life in the routing repo. What was missing here was the correct wire-up of values.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs [hunk 4]
```diff
@@ -133,24 +138,44 @@ private TreeRouter BuildRoute(ActionDescriptorCollection actions)
             var distinctRouteInfosByGroup = GroupRouteInfosByGroupId(routeInfos);
             foreach (var routeInfo in distinctRouteInfosByGroup)
             {
-                routeBuilder.Add(new TreeRouteMatchingEntry()
+                // Note that because we only support 'inline' defaults, each routeInfo group also has the same
+                // set of defaults. 
+                //
+                // We then inject the route group as a default for the matcher so it gets passed back to MVC
+                // for use in action selection.
+                var defaults = new RouteValueDictionary(routeInfo.Defaults);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs]: Does this need a `StringComparer.OrdinalIgnoreCase` comparer?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs]: RVD is always `OrdinalIgnoreCase `

**Comment by pranavkm** [comment]
:shipit:

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/AttributeRouteTest.cs [hunk 5]
```diff
@@ -118,5 +103,513 @@ public class AttributeRouteTest
...
+                    Assert.Empty(e.Constraints);
+                    Assert.Empty(e.Defaults);
+                    Assert.Equal(RoutePrecedence.ComputeGenerated(e.Template), e.GenerationPrecedence);
+                    Assert.Equal("BLOG_INDEX", e.Name);
+                    Assert.Equal(17, e.Order);
+                    Assert.Equal(actions[0].RouteValueDefaults, e.RequiredLinkValues);
+                    Assert.Equal("1", e.RouteGroup);
+                    Assert.Equal("api/Blog/{id}", e.Template.TemplateText);
+                },
+                e =>
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/AttributeRouteTest.cs]: Is this repeated?

File: test/WebSites/RoutingWebSite/Controllers/TeamController.cs [hunk 6]
```diff
@@ -63,8 +63,8 @@ public ActionResult GetAllTeams(int notRelevant)
             return Content(Url.Action(), "text/plain");
         }
 
-        [HttpGet("/TeamName/{*Name}/")]
-        public ActionResult GetTeam(string name = "DefaultName")
```
Comment (javiercn) [inline test/WebSites/RoutingWebSite/Controllers/TeamController.cs]: This is not supported?

**Comment by javiercn** [comment]
:shipit:

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/AttributeRouteTest.cs (hunk 5 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/AttributeRouteTest.cs]: There are supposed to be two entries.

File: test/WebSites/RoutingWebSite/Controllers/TeamController.cs (hunk 6 above)
Comment (rynowak) [inline test/WebSites/RoutingWebSite/Controllers/TeamController.cs]: It won't be in RTM

Comment (rynowak) [inline test/WebSites/RoutingWebSite/Controllers/TeamController.cs]: This is inconsistent with how conventional routes work with catchalls and inconsistent with our model binding works.

**Pull request merged**